Add new `ex history` command

jlebon opened the pull request:
This is the rpm-ostree equivalent of `dnf history`. As opposed to the
history of the refspec (i.e. `ostree log`), this shows the history of
the system, i.e. the refspecs the host deployed, checksums, versions,
layered packages, etc... The amount of details remembered is similar to
what shows up in `status`.

There's definitely some further enhancements possible (e.g. printing
package diffs, displaying rollbacks), though this seems in good enough
shape as a first cut.

Closes: #1489

jlebon commented:
Requires https://github.com/ostreedev/ostree/pull/1842

cgwalters commented on src/app/rpmostree-ex-builtin-history.c:
```diff
@@ -0,0 +1,52 @@
+                                       error))
+    return FALSE;
+
+  g_print ("hello world\n");
```
This one is defunct?

cgwalters commented on rust/src/history.rs:
```diff
@@ -0,0 +1,733 @@
+
+static RPMOSTREE_UPGRADER_DATA_DIR: &'static str = "/var/lib/rpm-ostree/history";
+
+pub struct HistoryCtx {
```
Be nice to have at least one-liner rustdoc comments for structs like this.

cgwalters commented on rust/src/history.rs:
```diff
@@ -0,0 +1,733 @@
+    reached_eof: bool,
+}
+
+// This whole struct is exposed to the C side.
```
`/// A single history entry exposed via FFI to C``` ?

jlebon commented:
Fixup! :arrow_up: 

I added docstrings on the functions that missed them and a header describing the overall approach. In the process, I ended up renaming a bunch of things to clearer, more appropriate, names.

jlebon commented:
Rebased! (Just marking as "Ready for review" to pass it through CI).

jlebon commented:
bot, retest this please

lucab commented:
One overall doubt that the docstring didn't answer: is there some garbage-collection mechanism to be considered/planned/implemented here for history files?

miabbott commented:
/me bikesheds

In a recent demo of this work, @jlebon pointed out that `DeployCommand` was indicating the command that was responsible for creating the deployment.  So if a user used `rpm-ostree rollback` to go back to the previous deployment, the `DeployCommand` wouldn't be `rollback`, but rather `install ltrace` or `override remove podman`.

@dustymabe suggested that we go a step further and track this difference, so `rollback` would be recorded in the history somehow.

I think this would be excellent, as long as we could display this info in a sane way.  Maybe use `DeployCommand` in the case of a `rollback`, but also have something like `DeployCreateCommand` to show the original command used?   

The more I think on this, it seems like it's really only the `rollback` command where confusion would arise.  Other operations (`deploy`, `install`, `rebase`, `upgrade`, etc) are really about creating a brand new deployment, where as `rollback` is about using a previously created deployment and you would want to know how that deployment was created.

cgwalters commented on rust/src/history.rs:
```diff
@@ -0,0 +1,783 @@
+            Some(ftype) => ftype,
+            /* stat() fallback */
+            None => dir.metadata(entry.file_name())?.simple_type(),
+        };
```
This is also `get_file_type()` in `OpenatDirExt`.  (But eh, don't need to port to openat if you don't want to)

cgwalters commented on src/daemon/rpmostree-sysroot-upgrader.c:
```diff
@@ -1226,6 +1230,74 @@ rpmostree_sysroot_upgrader_set_kargs (RpmOstreeSysrootUpgrader *self,
+   * it breaks the expectation that journal messages should be somewhat easily
+   * introspectable. We could also serialize it to JSON first, though we wouldn't be able to
+   * re-use the printing code in `status.c` as is. Note also the GVariant can be large (e.g.
+   * we include the full `rpmostree.rpmdb.pkglist` in there). */
```
What elements of this deployment variant do we use?  The other option is to log those as structured fields right?

jlebon replied:
> What elements of this deployment variant do we use?

Basically almost everything. :) We pass it to the same function in `status.c` that normally prints it. I also want to make use of the pkglist in the future so that we can show the diff from one version to the next.

>  The other option is to log those as structured fields right?

Yeah, I had briefly looked at this. One obvious annoyance is that the field naming convention means we have to do some translation back and forth for the top-level keys. And the values themselves I guess would have to be run through `g_variant_print` and `g_variant_parse`. Still, I go back to whether we should shove that much data in the journal in the first place. (Even though the actual limit is a crazy amount like 768M...). It somewhat reminds me of the D-Bus/`--json` discussion around pkglists and how we ended up blacklisting it, even though it's *mostly* for machines.

jlebon commented:
> We're a bit lacking real e2e tests

Yeah, I'm pretty confident in the unit tests, though I definitely intend to add some basic vmcheck tests as well!

jlebon commented:
I just realized now that I squashed changes into the commit and rebased in one shot, which makes it hard to see what changed. Anyway, it was mostly a rebase + supporting `--json` + using `get_file_type()`. Still need to add tests and address some of the feedback from https://github.com/projectatomic/rpm-ostree/pull/1813#issuecomment-490183965 (though yeah... might leave that as a follow-up).

jlebon commented:
Now with tests! Dropped WIP and updated the original comment as well with the final commit message.

I haven't addressed the feedback in https://github.com/projectatomic/rpm-ostree/pull/1813#issuecomment-490183965 other than renaming the key names to `CreateCommand` and `CreateTimestamp`. Will open a follow-up ticket for the remaining items.

jlebon commented:
:heavy_check_mark: Tests all green now!
Let's ship this?

cgwalters commented:
I think this is safe enough to try out and use to gather feedback.  Thanks for the work on this!

@rh-atomic-bot r+ f49100a

jlebon commented:
@rh-atomic-bot retry

jlebon commented:
@rh-atomic-bot retry

jlebon commented:
@rh-atomic-bot retry

jlebon commented:
@rh-atomic-bot retry
